example_buffer.py

```diff
@@ -0,0 +1,198 @@
+        _rsync_dir(os.path.join(
+            source_dir, m[1]), os.path.join(dest_dir, m[1]))
+
+
+def _rsync_dir(source_dir, dest_dir):
+    _ensure_dir_exists(dest_dir)
+    with open('.rsync_log', 'ab') as rsync_log:
+        subprocess.call(['gsutil', '-m', 'rsync', source_dir, dest_dir],
+                        stderr=rsync_log)
+
+
+def fill_and_wait(bufsize=dual_net.EXAMPLES_PER_GENERATION,
+                  write_dir=rl_loop.GOLDEN_CHUNK_DIR,
+                  model_window=100,
+                  threads=8,
+                  skip_first_rsync=False):
+    buf = ExampleBuffer(bufsize)
+    models = rl_loop.get_models()[-model_window:]
+    # Last model is N.  N+1 is training.  We should gather games for N+2.
+    chunk_to_make = os.path.join(rl_loop.GOLDEN_CHUNK_DIR, str(
+        models[-1][0] + 2) + '.tfrecord.zz')
+    while tf.gfile.Exists(chunk_to_make):
+        print("Next chunk ({}) already exists.  Sleeping.".format(chunk_to_make))
+        time.sleep(5*60)
+        models = rl_loop.get_models()[-model_window:]
+    print("Making chunk:", chunk_to_make)
+    if not skip_first_rsync:
+        with timer("Rsync"):
+            smart_rsync(models[-1][0] - 6)
+    files = list(tqdm(map(files_for_model, models), total=len(models)))
+    buf.parallel_fill(list(itertools.chain(*files)), threads=threads)
+
+    print("Filled buffer, watching for new games")
+    while rl_loop.get_latest_model()[0] == models[-1][0]:
+        with timer("Rsync"):
+            smart_rsync(models[-1][0] - 2)
+        new_files = list(
+            tqdm(map(files_for_model, models[-2:]), total=len(models)))
+        buf.update(list(itertools.chain(*new_files)))
+        time.sleep(60)
+    latest = rl_loop.get_latest_model()
+
+    print("New model!", latest[1], "!=", models[-1][1])
+    print(buf)
+    buf.flush(os.path.join(write_dir, str(latest[0]+1) + '.tfrecord.zz'))
+
+
+def make_chunk_for(output_dir=LOCAL_DIR,
+                   game_dir=rl_loop.SELFPLAY_DIR,
+                   model_num=1,
+                   positions=dual_net.EXAMPLES_PER_GENERATION,
+                   threads=8,
+                   samples_per_game=4):
+    models = {k: v for k, v in rl_loop.get_models()}
+    buf = ExampleBuffer(positions)
+    cur_model = model_num - 1
+    files = []
+    while (len(files) * samples_per_game) < positions and cur_model >= 0:
+        local_model_dir = os.path.join(LOCAL_DIR, models[cur_model])
+        if not tf.gfile.Exists(local_model_dir):
+            print("Rsyncing", models[cur_model])
+            _rsync_dir(os.path.join(
+                game_dir, models[cur_model]), local_model_dir)
+        files += tf.gfile.Glob(os.path.join(local_model_dir, '*.zz'))
+        cur_model -= 1
+
+    print("Filling from {} files".format(len(files)))
+
+    buf.parallel_fill(files, threads=threads,
+                      samples_per_game=samples_per_game)
+    print(buf)
+    output = os.path.join(output_dir, str(model_num) + '.tfrecord.zz')
```

Comment:
same question here as above - this is a local file; how does trainer get access to this file?

Reply:
we write out gs:// path, but rsync locally for faster random reads on small files.